src/merkle_tree.rs

```diff
@@ -0,0 +1,282 @@
+            let transaction_hash = hasher.finish();
+            transactions_hash.push(transaction_hash);
+        }
+        transactions_hash
+    }
+
+    pub fn verify(&mut self, transaction: H, proof: Vec<SiblingHash>) -> bool {
+        let mut hasher = DefaultHasher::new();
+        transaction.hash(&mut hasher);
+        let mut transaction = hasher.finish();
+        for sibling_hash in proof {
+            hasher = DefaultHasher::new();
+            match sibling_hash {
+                SiblingHash::Left(left_hash) => {
+                    left_hash.hash(&mut hasher);
+                    transaction.hash(&mut hasher);
+                }
+                SiblingHash::Right(right_hash) => {
+                    transaction.hash(&mut hasher);
+                    right_hash.hash(&mut hasher);
+                }
+            }
+
+            transaction = hasher.finish();
+        }
+
+        transaction == self.merkle_root.hash_value
+    }
+
+    fn recursive_get_proof(
+        current_node: &MerkleNode,
+        proof: &mut Vec<SiblingHash>,
+        transaction_hash: u64,
+    ) -> bool {
+        if let Some(left) = &current_node.left_son {
+            if left.hash_value == transaction_hash {
+                if let Some(right_sibling) = &current_node.right_son {
+                    proof.push(SiblingHash::Right(right_sibling.hash_value));
+                }
+                return true;
+            }
+            if Self::recursive_get_proof(left, proof, transaction_hash) {
+                if let Some(right_sibling) = &current_node.right_son {
+                    proof.push(SiblingHash::Right(right_sibling.hash_value));
+                }
+                return true;
+            }
+        }
+
+        if let Some(right) = &current_node.right_son {
+            if right.hash_value == transaction_hash {
+                if let Some(left_sibling) = &current_node.left_son {
+                    proof.push(SiblingHash::Left(left_sibling.hash_value));
+                }
+                return true;
+            }
+
+            if Self::recursive_get_proof(right, proof, transaction_hash) {
+                if let Some(left_sibling) = &current_node.left_son {
+                    proof.push(SiblingHash::Left(left_sibling.hash_value));
+                }
+                return true;
+            }
+        }
+        false
+    }
+
+    pub fn get_proof(&mut self, transaction: H) -> Vec<SiblingHash> {
+        let mut proof = Vec::new();
+        let mut hasher = DefaultHasher::new();
+        transaction.hash(&mut hasher);
+        Self::recursive_get_proof(&self.merkle_root, &mut proof, hasher.finish());
+
+        proof
+    }
+
+    pub fn add(&mut self, transaction: H) {
+        self.leafs.push(transaction);
+        let mut leafs = Vec::new();
+        for leaf in &self.leafs {
+            leafs.push(leaf.clone());
+        }
+        self.merkle_root = Self::create_tree(leafs.clone()).merkle_root;
```

Comment:
```suggestion
        self.merkle_root = Self::create_tree(self.leafs.clone()).merkle_root;
```
The function is ok but this works too.